From 615aa658683b4f61e211cf3f4e896fde55c523b6 Mon Sep 17 00:00:00 2001
From: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
Date: Thu, 20 Mar 2025 14:18:14 +0800
Subject: [PATCH 184/236] mtk: hostapd: Add support for printing eml_cap using
 hostapd_cli

Add support for printing emlsr_support, emlmr_support, nstr_bitmap
using hostapd_cli sta command.

Usage:
hostapd_cli -i <interface> -l <link> sta <MLD addr>

e.g.
hostapd_cli -i ap-mld-1 -l 0 sta 06:0c:43:2a:ee:f3
 ...
 peer_addr[0]=06:0c:43:2a:ee:f2
 peer_addr[1]=06:0c:43:2a:ee:f3
 nstr_bitmap=0x05
 peer_addr[2]=06:0c:43:2a:ee:f4
 nstr_bitmap=0x03
 max_simul_links=2
 emlsr_support=0
 emlmr_support=0

Signed-off-by: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
---
 src/ap/ctrl_iface_ap.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/src/ap/ctrl_iface_ap.c b/src/ap/ctrl_iface_ap.c
index d315eeba7..2328d42e9 100644
--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -477,10 +477,17 @@ static int hostapd_ctrl_iface_sta_mib(struct hostapd_data *hapd,
 #ifdef CONFIG_IEEE80211BE
 	if (sta->mld_info.mld_sta) {
 		u16 mld_sta_capa = sta->mld_info.common_info.mld_capa;
+		u16 mld_sta_eml_capa = sta->mld_info.common_info.eml_capa;
 		u8 max_simul_links = mld_sta_capa &
 			EHT_ML_MLD_CAPA_MAX_NUM_SIM_LINKS_MASK;
+		u8 emlsr_support = mld_sta_eml_capa &
+			EHT_ML_EML_CAPA_EMLSR_SUPP;
+		u8 emlmr_support = mld_sta_eml_capa &
+			EHT_ML_EML_CAPA_EMLMR_SUPP;
 
 		for (i = 0; i < MAX_NUM_MLD_LINKS; ++i) {
+			int j;
+
 			if (!sta->mld_info.links[i].valid)
 				continue;
 			ret = os_snprintf(
@@ -489,12 +496,31 @@ static int hostapd_ctrl_iface_sta_mib(struct hostapd_data *hapd,
 				i, MAC2STR(sta->mld_info.links[i].peer_addr));
 			if (!os_snprintf_error(buflen - len, ret))
 				len += ret;
+
+			for (j = 0; j < sta->mld_info.links[i].nstr_bitmap_len; j++) {
+				ret = os_snprintf(buf + len, buflen - len,
+						  "nstr_bitmap=0x%02x\n",
+						  sta->mld_info.links[i].nstr_bitmap[j]);
+				if (!os_snprintf_error(buflen - len, ret))
+					len += ret;
+			}
+
 		}
 
 		ret = os_snprintf(buf + len, buflen - len,
 				  "max_simul_links=%d\n", max_simul_links);
 		if (!os_snprintf_error(buflen - len, ret))
 			len += ret;
+
+		ret = os_snprintf(buf + len, buflen - len,
+				  "emlsr_support=%d\n", emlsr_support);
+		if (!os_snprintf_error(buflen - len, ret))
+			len += ret;
+
+		ret = os_snprintf(buf + len, buflen - len,
+				  "emlmr_support=%d\n", emlmr_support);
+		if (!os_snprintf_error(buflen - len, ret))
+			len += ret;
 	}
 #endif /* CONFIG_IEEE80211BE */
 
-- 
2.45.2

